ext/smtp/sink.go

```diff
@@ -299,3 +676,58 @@ func hashMetadata(em emailMetadata) string {
 func getAttachmentPath(em emailMetadata, attachment string) string {
 	return filepath.Join(tmpFolder, hashMetadata(em), attachment)
 }
+
+// TODO need to ensure that TMP path has hash metadata in it
+func getTmpPath(destinationURI string) (string, error) {
+	targetURI, err := url.Parse(destinationURI)
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+	return filepath.Join("/tmp", targetURI.Path), nil
+}
+
+func getOSSPath(em emailMetadata, attachment string, ossDestinationDir string) string {
+	return fmt.Sprintf("%s/%s/%s", strings.TrimRight(ossDestinationDir, "/"), hashMetadata(em), attachment)
+}
+
+func (s *SMTPSink) flushToOSS(destinationURI string, oh io.WriteCloser) error {
+	// open tmp file for reading
+	tmpPath, err := getTmpPath(destinationURI)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	var tmpReader io.ReadSeekCloser
+	tmpReader, err = os.OpenFile(tmpPath, os.O_RDONLY, 0644)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	// header is skipped if SKIP_HEADER is explicitly set to true OR if file has been partially uploaded previously
+	skipHeader := s.skipHeader || (s.maxTempFileRecordNumber > 0 && s.fileRecordCounters[tmpPath] > s.maxTempFileRecordNumber)
+
+	// convert to appropriate format if necessary
+	switch filepath.Ext(destinationURI) {
+	case ".json":
+		// do nothing
+	case ".csv":
+		tmpReader = helper.FromJSONToCSV(s.Logger(), tmpReader, skipHeader)
+	case ".tsv":
+		tmpReader = helper.FromJSONToCSV(s.Logger(), tmpReader, skipHeader, rune('\t'))
+	default:
```

Comment:
let's add one more extension, for .xlsx as it's already support in the latest changes main branch